### PR TITLE
[Bugfix] Fix location of `unscale` in mixed precision plugin

### LINKED ISSUE #9599: Clip norm after scaler.unscale_ in native fp16 training
## 🐛 Bug
I trained a large model using native amp, but the loss converged very slow. After a careful check of the backward and optimization code, I found the `clip_gradients` is executed right after backward, but `scaler.unscale_` is conducted in pre_optimization_step.
According to the [instruction of Pytorch](https://pytorch.org/docs/stable/notes/amp_examples.html#gradient-clipping), the order of clip and unscale should be exchanged. Currently `gradient_clip_val` may lead to a very flat learning curve if used together with native amp.
Hope to be fixed.

### PR DESCRIPTION
**Pull request by justusschock**
## What does this PR do?

Fixes #9599, #9330, #9205, #9694

Unscale was called after gradnorm tracking and gradient clipping. Moving it to a different hook in the plugins should fix that.

Alternative solution would be to move gradient tracking and clipping. Finding a solution that works for all plugin combinations might be difficult though.

Probably same purpose as #9287

### Does your PR introduce any breaking changes? If yes, please list them.

`None`

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

**Please don't merge before review from @carmocca**

### CODE REVIEW

File: pytorch_lightning/plugins/precision/native_amp.py [hunk 1]
```diff
@@ -106,6 +105,15 @@ def pre_optimizer_step(
             self.scaler.update()
         return False
 
+    def post_backward(
+        self, model: "pl.LightningModule", closure_loss: torch.Tensor, optimizer: torch.optim.Optimizer
+    ) -> torch.Tensor:
+        ret_val = super().post_backward(model, closure_loss, optimizer)
+        # unscale here to have it inside the closure before the grad tracking and clipping
+        if not model.trainer.fit_loop._should_accumulate():
+            self.scaler.unscale_(optimizer)
```
Comment (tchaton) [inline pytorch_lightning/plugins/precision/native_amp.py:114]: Can you check this isn't the LightningOptimizer ?

File: tests/plugins/test_amp_plugins.py [hunk 2]
```diff
@@ -232,3 +232,39 @@ def test_cpu_amp_precision_16_throws_error(tmpdir):
...
+    def post_backward(
+        self, model: "pl.LightningModule", closure_loss: torch.Tensor, optimizer: torch.optim.Optimizer
+    ) -> torch.Tensor:
+        norm_before = torch.nn.utils.clip_grad_norm_(model.parameters(), 2)
+        ret_val = super().post_backward(model, closure_loss, optimizer)
+        norm_after = torch.nn.utils.clip_grad_norm_(model.parameters(), 2)
+
+        # norm_after unscale should be smaller by scaling factor greater than 1
+        if not (torch.isinf(norm_before) or torch.isnan(norm_before)):
+            assert norm_after < norm_before
```
Comment (tchaton) [inline tests/plugins/test_amp_plugins.py:249]: The scale should be much larger than 1. 

```suggestion
            assert norm_after < norm_before * 10
```

**Comment by awaelchli** [comment]
@justusschock I think this is ready for review now

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/plugins/precision/native_amp.py [hunk 3]
```diff
@@ -96,7 +96,8 @@ def pre_optimizer_step(
                 " To request, please file a Github issue in PyTorch and tag @mcarilli"
             )
         result = lambda_closure()  # native amp does not support closures
-        self.scaler.unscale_(optimizer)
+        if not model.automatic_optimization:
```
Comment (carmocca) [inline pytorch_lightning/plugins/precision/native_amp.py:99]: Why is this if added? `step` below will not unscale again if it has been done manually already

File: pytorch_lightning/accelerators/accelerator.py [hunk 4]
```diff
@@ -228,7 +228,7 @@ def validation_step_end(self, output: Optional[STEP_OUTPUT]) -> Optional[STEP_OU
         """
         return self.training_type_plugin.validation_step_end(output)
 
-    def backward(self, closure_loss: Tensor, *args: Any, **kwargs: Any) -> Tensor:
+    def backward(self, closure_loss: Tensor, optimizer: torch.optim.Optimizer, *args: Any, **kwargs: Any) -> Tensor:
```
Comment (carmocca) [inline pytorch_lightning/accelerators/accelerator.py:231]: as a side note: this new parameter would be removed again with the long-term solution of moving the `step` into the precision plugin

Comment (four4fish) [inline pytorch_lightning/accelerators/accelerator.py:231]: Does Optimizer vs torch.optim.Optimizer the same here?

File: pytorch_lightning/plugins/precision/native_amp.py (hunk 3 above)
Comment (four4fish) [inline pytorch_lightning/plugins/precision/native_amp.py:99]: Do we need check isinstance(model, pl.LightningModule) here? Is it possible that model is wrapped ?

Comment (awaelchli) [reply pytorch_lightning/plugins/precision/native_amp.py:99]: @carmocca I don't like it either. This makes the hacks that exist here even more hacky. 
The PR aims at solving the bug that unscaling is happening after gradient clipping. So to fix this, one can move the unscaling to an earlier point, which we choose to do directly after backward (post_backward). However, we ONLY want to do that in the case of automatic optimization, because gradient clipping does not exist for manual backward and we need to keep the unscaling in the optimizer_step here because the user of manual optimization will call the LightningOptimizer.step themselves. 

Bottom line: we have unscale in two locations, but only one is ever active depending on whether we are in automatic or manual optimization.

File: pytorch_lightning/accelerators/accelerator.py (hunk 4 above)
Comment (awaelchli) [reply pytorch_lightning/accelerators/accelerator.py:231]: yes, we should choose just "Optimizer" here for consistency with the other signatures. It does not make a functional difference :)

File: pytorch_lightning/plugins/precision/native_amp.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/precision/native_amp.py:99]: No, not needed. 1) the methods here are type annotated with LightningModule. 2) we are explicitly passing in `self.lighting_module` where we call the method, and `self.lightning_module` is guaranteed to be the unwrapped pure LM.

File: pytorch_lightning/plugins/precision/native_amp.py [hunk 5]
```diff
@@ -96,7 +96,8 @@ def pre_optimizer_step(
...
+            self.scaler.unscale_(optimizer)
```
Comment (SeanNaren) [inline pytorch_lightning/plugins/precision/native_amp.py:100]:
```suggestion
            # unscale in manual optimization as user does not rely on lightning
            # to call backward, but does call LightningOptimizer.step
            self.scaler.unscale_(optimizer)
```

**Review by SeanNaren — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

**Comment by awaelchli** [comment]
As discussed with @carmocca I rebased this onto the 1.4.x bugfix branch as we desire a better, less hacky long-term solution for master #9287 . So this is considered a hotfix.

There were basically no merge conflicts, the changes stay the same.

File: tests/plugins/test_amp_plugins.py (hunk 2 above)
Comment (awaelchli) [reply tests/plugins/test_amp_plugins.py:189]: I don't fully understand, adding the `* 10` on the right hand side of the inequality is not going to add any more value to the assertion

File: pytorch_lightning/plugins/precision/native_amp.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/precision/native_amp.py:77]: done, added assertion

File: tests/plugins/test_amp_plugins.py (hunk 2 above)
Comment (tchaton) [reply tests/plugins/test_amp_plugins.py:191]: The amp scaling value is quite large. Could it be possible the norm after is smaller due to un-precision `norm_after < norm_before`.

The *10 would just prevent this type of regression. This is optional.

**Review by tchaton — APPROVED** [review]
LGTM !
